### DISCUSSION TITLE
How to add .jar and .aar to the native module

### DISCUSSION THREAD

**Discussion by roni-castro-glooko** [callstack/react-native-builder-bob#397] [Q&A]
I have some .jar and .aar that needs to be integrated in a native module to be used as `import`. I tried adding them to the top-level Gradle file (android/build.gradle), but they aren't being added to the example app as a dependency.  How can I do it?

What I tried:

```js
// added the jars/aar to the `android/libs` folder:
android/libs/android-transmitter-4.1.10.aar
android/libs/android-transmitter-core-4.1.10.jar
```

## Solution 1

```js
// android/build.gradle
repositories {
  mavenCentral()
  google()
  flatDir {
    dirs 'libs'
  }
}

dependencies {
  // added these to the top
  implementation fileTree(include: ['*.jar','*.aar'], dir: 'libs')
}
```

## Solution 2

```js
// android/build.gradle
repositories {
  mavenCentral()
  google()
  flatDir {
    dirs 'libs'
  }
}

dependencies {
    // added these to the top
    implementation fileTree(include: ['*.jar','*.aar'], dir: 'libs')
    implementation(name: 'android-transmitter', version: "4.1.10", ext: 'aar')
    implementation(name: 'android-transmitter-core', version: "4.1.10")
}
```

## Solution 3

```js
// android/build.gradle
dependencies {
  // ... others `implementation`
  implementation fileTree(include: ['*.jar','*.aar'], dir: 'libs')
}
```

## Solution 4

```js
// android/build.gradle
dependencies {
  implementation fileTree(include: ['*.jar','*.aar'], dir: 'libs')
 // ... others `implementation`
}
```

**Comment by roni-castro-glooko** [comment]
`Solution 3` or `Solution 4` works actually. 
The problem I had was basically because I gave the same package name of module that is being used by the .jar and .aar libs, so they were having conflicts when importing 😆 

```js
// `android/libs` folder:
android-transmitter-4.1.10.aar
android-transmitter-core-4.1.10.jar
```

```js
// android/build.gradle
dependencies {
  implementation fileTree(include: ['*.jar','*.aar'], dir: 'libs')
 // ... others `implementation`
}
```

**Comment by atlj** [comment]
Solution 3 and 4 might work for development builds but can you try a production build of the example app 😅 ?

**Reply by roni-castro-glooko** [reply]
I deployed the lib to Github packages, installed this lib in the production app, and i worked without any crash, if I generate an android apk or .aab(and convert back to .apk with bundletool) of the final app through fastlane, shouldn't it work? 🤔 

the android folder of the lib node_mobules installed in the project has all the libs, after the build with bob builder + deployed
![Screen Shot 2023-06-20 at 23 40 11](https://github.com/callstack/react-native-builder-bob/assets/131805209/207c47e2-efe6-4bef-a07f-c1ee9da962d6)

**Reply by atlj** [reply]
Could you try to get a release build by going to example/android and running `gradlew assembleRelease`? We have had problems with adding aar files like this in the past. There are two quick solutions to this:
1. Using fataar
2. Putting your aar files into a folder inside your project and treating that folder as a private local maven repository.

This problem is happening because of the transient dependencies. Any dependencies you have in your aar file doesnt get picked up by gradle resulting issues in build steps.

**Reply by Noitham** [reply]
Hello @atlj , is there any "official" solution for this issue?

I've tried using fataar, but didn't work either.

As you mentioned, putting the `implementation files('libs/... .aar')` does work when building the app locally, both staging and release variants, but as soon as we do `assembleRelease`, we get this error.

> Direct local .aar file dependencies are not supported when building an AAR. The resulting AAR would be broken because the classes and Android resources from any local .aar file dependencies would not be packaged in the resulting AAR. Previous versions of the Android Gradle Plugin produce broken AARs in this case too (despite not throwing this error).

**Reply by roni-castro-glooko** [reply]
I tried fataar and It didn't work for me. Also this dependency is not being updated anymore.

What worked for me is to couple the app with the library dependencies:

**In the library**

1.  Add the .aar to `android/libs`

![Screen Shot 2023-11-29 at 20 35 26](https://github.com/callstack/react-native-builder-bob/assets/131805209/761a5fb0-2cb2-415d-8f92-a2c5e38e1731)

2.  Inside the library `android/build.gradle` add `compileOnly fileTree(include: ['*.jar','*.aar'], dir: 'libs')`

```js
dependencies {
  compileOnly fileTree(include: ['*.jar','*.aar'], dir: 'libs') // Add compileOnly, so these files will not affect the build

  implementation "com.facebook.react:react-native:+"
}

```

--- 
**Inside your app** 

1. Inside your app `android/app/build.gradle`, which is installing the lib above, add this to get the .aar inside the lib folder

```js
implementation fileTree(include: ['*.jar','*.aar'], dir: '../../node_modules/my-react-native-library-name/android/libs')
```

2. For the example app, you can use the path to the libs folder, as react-native-bob-builder does not install the lib inside the node_modules

```js
implementation fileTree(include: ['*.jar','*.aar'], dir: '../../../android/libs')
```
